### PR TITLE
adding checks for json end_document in http transport

### PR DESCRIPTION
**Pull request by duzinkie**
the issue is discussed in: https://github.com/apollographql/apollo-kotlin/issues/5885

this is the recommended fix for the no http batching scenario, for the main branch

### CODE REVIEW

File: libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/internal/ResponseParser.kt [hunk 1]
```diff
@@ -45,6 +45,10 @@ internal object ResponseParser {
 
     jsonReader.endObject()
 
+    if (jsonReader.peek() != JsonReader.Token.END_DOCUMENT) {
+      println("Apollo: extra tokens after payload")
+    }
```
Comment (martinbonnin) [inline libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/internal/ResponseParser.kt:50]: 2 things:

1. Can you move this to the `toApolloResponse()` function? I'd like to keep the possibility to read a single JSON object from `ResponseParser.parse` in the future
2. Let's make it an error? For this you can `throw JsonDataException("Expected END_DOCUMENT but was ${peek()}")`

**Comment by martinbonnin** [comment]
Hi @duzinkie 👋 . Want to get this through the finish line?

**Comment by duzinkie** [comment]
Hey @martinbonnin - sorry for the delays, I was on vacation - will be catching up on this this week.

File: libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/internal/ResponseParser.kt (hunk 1 above)
Comment (duzinkie) [reply libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/internal/ResponseParser.kt]: I've made changes (hopefully) - as requested - lmk, and I'll push an equivalent to the 3.x branch.

Also - do you want the same applied to batching interceptor parsing - I'm guessing an exception should also be thrown there, not sure about whether code needs moving around?

Comment (martinbonnin) [reply libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/internal/ResponseParser.kt]: > I've made changes (hopefully) - as requested - lmk, and I'll push an equivalent to the 3.x branch.

Thanks, LGTM 👍 

> Also - do you want the same applied to batching interceptor parsing - I'm guessing an exception should also be thrown there, not sure about whether code needs moving around?

Yup, same exception, no need to move code around.

Comment (duzinkie) [reply libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/internal/ResponseParser.kt]: ok, Ive pushed a similar change to the other 3 prs.
not sure whether I should investigate integration test failures here?

Comment (martinbonnin) [reply libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/internal/ResponseParser.kt]: > ok, Ive pushed a similar change to the other 3 prs.

Thanks!

> not sure whether I should investigate integration test failures here?

Doesn't look related, I'll take a deeper look tomorrow

Comment (martinbonnin) [reply libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/internal/ResponseParser.kt]: Turns out it was related. `peek()` can throw but `toApolloResponse()` shouldn't. See https://github.com/apollographql/apollo-kotlin/pull/5895/commits/9ab202cc3536b505a8387f0e619db3660cf2ba62

I'll take it from there for this PR and the other 3. Thanks again for looking into this!

**Review by BoD — APPROVED** [review]

**Pull request merged**